### PR TITLE
PP-1156 Job Obit use memory after free

### PR DESCRIPTION
**Pull request by vstumpf**
#### Issue-ID

* **[PP-1156](https://pbspro.atlassian.net/browse/PP-1156)**

#### Problem description
* req_jobobit.c:rel_resc frees the batch_request, then accesses it. Crashes with ASAN

#### Cause / Analysis
* Dont use memory after freeing.

#### Solution description
* Saved a pointer to the connection, free the br, then access the connection via the pointer.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/server/req_jobobit.c [hunk 1]
```diff
@@ -633,14 +633,15 @@ rel_resc(job *pjob)
 	/* is there a rerun request waiting for acknowledgement that        */
 	/* resources (including licenses) are indeed released? Then ack it. */
 	if (pjob->ji_rerun_preq != NULL) { /* set only in req_rerun() */
+		/* save pointer to connection */
+		if(pjob->ji_rerun_preq->rq_conn != PBS_LOCAL_CONNECTION)
```
Comment (bhroam) [inline src/server/req_jobobit.c]: Add a space after the if

File: src/server/req_jobobit.c [hunk 2]
```diff
@@ -633,14 +633,15 @@ rel_resc(job *pjob)
 	/* is there a rerun request waiting for acknowledgement that        */
 	/* resources (including licenses) are indeed released? Then ack it. */
 	if (pjob->ji_rerun_preq != NULL) { /* set only in req_rerun() */
+		/* save pointer to connection */
```
Comment (bhroam) [inline src/server/req_jobobit.c]: I don't think this comment is necessary

File: src/server/req_jobobit.c [hunk 3]
```diff
@@ -633,14 +633,14 @@ rel_resc(job *pjob)
...
+
 		reply_ack(pjob->ji_rerun_preq);
 
 		/* clear no-timeout flag on connection */
-		if (pjob->ji_rerun_preq->rq_conn != PBS_LOCAL_CONNECTION) {
-			conn = get_conn(pjob->ji_rerun_preq->rq_conn);
-			if (conn)
-				conn->cn_authen &= ~PBS_NET_CONN_NOTIMEOUT;
-		}
+		if (conn)
```
Comment (agrawalravi90) [inline src/server/req_jobobit.c]: I know it's not your comment, but could you please expand this and explain why this flag needs to be cleared? Might save people some time. But it's up to you, I won't hold your check-in for this. Thanks

**Comment by mike0042** [comment]
Seeing the following on Debian...

failed: test_resource_delete (tests.pbs_smoketest.SmokeTest)
Test cases with failures: SmokeTest.test_resource_delete
Test suites with failures: SmokeTest
run: 49, succeeded: 48, failed: 1, errors: 0, skipped: 0, timedout: 0

Restarted the build and produced the same failure.

**Comment by agrawalravi90** [comment]
This test fails intermittently, i think it's  a bug in the test itself.

- Ravi

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by mike0042** [comment]
I just kicked off the Debian build for a third time. If it fails again, the problem is more than intermittent and needs to be addressed (in a separate PR).

**Comment by agrawalravi90** [comment]
It passed this time, I logged a bug for it here: https://pbspro.atlassian.net/browse/PP-1163

**Comment by vstumpf** [comment]
Thanks!

**Pull request merged**